library/core/src/async_iter/async_iter.rs

```diff
@@ -1,54 +1,27 @@
-use crate::ops::DerefMut;
-use crate::pin::Pin;
-use crate::task::{Context, Poll};
-
 /// A trait for dealing with asynchronous iterators.
 ///
+/// This trait is an alternative to the [`Stream`] trait. Both traits are
+/// currently being evaluated on nightly with the intent to stabilize only one.
+///
 /// This is the main async iterator trait. For more about the concept of async iterators
 /// generally, please see the [module-level documentation]. In particular, you
 /// may want to know how to [implement `AsyncIterator`][impl].
 ///
 /// [module-level documentation]: index.html
 /// [impl]: index.html#implementing-async-iterator
+/// [`Stream`]: crate::stream::Stream
 #[unstable(feature = "async_iterator", issue = "79024")]
 #[must_use = "async iterators do nothing unless polled"]
 #[doc(alias = "Stream")]
-#[lang = "async_iterator"]
+#[allow(async_fn_in_trait)]
 pub trait AsyncIterator {
     /// The type of items yielded by the async iterator.
     type Item;
 
     /// Attempt to pull out the next value of this async iterator, registering the
     /// current task for wakeup if the value is not yet available, and returning
     /// `None` if the async iterator is exhausted.
-    ///
-    /// # Return value
-    ///
-    /// There are several possible return values, each indicating a distinct
-    /// async iterator state:
-    ///
-    /// - `Poll::Pending` means that this async iterator's next value is not ready
-    /// yet. Implementations will ensure that the current task will be notified
-    /// when the next value may be ready.
-    ///
-    /// - `Poll::Ready(Some(val))` means that the async iterator has successfully
-    /// produced a value, `val`, and may produce further values on subsequent
-    /// `poll_next` calls.
-    ///
-    /// - `Poll::Ready(None)` means that the async iterator has terminated, and
-    /// `poll_next` should not be invoked again.
-    ///
-    /// # Panics
-    ///
-    /// Once an async iterator has finished (returned `Ready(None)` from `poll_next`), calling its
-    /// `poll_next` method again may panic, block forever, or cause other kinds of
-    /// problems; the `AsyncIterator` trait places no requirements on the effects of
-    /// such a call. However, as the `poll_next` method is not marked `unsafe`,
-    /// Rust's usual rules apply: calls must never cause undefined behavior
-    /// (memory corruption, incorrect use of `unsafe` functions, or the like),
-    /// regardless of the async iterator's state.
-    #[cfg_attr(not(bootstrap), lang = "async_iterator_poll_next")]
-    fn poll_next(self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Option<Self::Item>>;
+    async fn next(&mut self) -> Option<Self::Item>;
```

Comment:
I'm concerning about the semantic of calling `next` before the previous returned `Future` being drained to completion. It is not documented clearly.

```rust
let _ = ait.next(); // May be non-trivial if it's implemented as RPIT.
assert_eq!(poll_once(pin!(ait.next())), Pending); // Or poll then drop.
let elem = ait.next().await;
// Is elem the first, second or third item?
// or it depends?
// or continuing the previous unfinished one?
// or it's unspecified and may panic?
```

Since this is safe, extra care must be taken by implementation for that `Self` must be consistent and resume-able (or interruption detectable) at any interruption point aka. `await` in `next`. It's somewhat like `UnwindSafe` but for await.

Reply:
`fn next` takes `&mut self`, which is captured by the returned future. As long as the first future is live, the Rust borrow checker ensures that additional calls to methods containing `&mut self` will result in compiler errors.

Reply:
I've covered this topic in some depth on my blog [here](https://blog.yoshuawuyts.com/async-iterator-trait/#cancellation-safety) and [here](https://blog.yoshuawuyts.com/futures-concurrency-3/#issues-with-futures-select).

To summarize: I believe the overwhelming majority of issues concerning cancellation + recreation of futures can be attributed to problematic concurrency primitives; and `select!` in particular. Relying on recreating futures as a primary means of control-flow not only has poor performance, it also relies on the documentation-only "cancellation-safety" feature. This is the same property you seem to be inquiring about.

The short answer to your question about what happens is: like with any trait using implementable async functions, that's up to the trait implementation to document. As opposed to with poll-based functions where it's up to the trait definition to document.

I don't believe it's a good idea to be relying on a documentation-only property for control-flow like this; especially if the repercussions of getting it wrong include data loss and data corruption. Instead we should be looking to either bring this property into the type system (which I don't have high hopes for), or make this issue uncommon by instead providing [concurrency primitives](https://docs.rs/futures-concurrency/latest/futures_concurrency/) which don't rely on docs-only properties for correctness.